git-credential-subpath

```diff
@@ -0,0 +1,192 @@
+#!/bin/bash
+#
+# A credential helper to look up username based on path components.
+#
+# git credentials are able to be configured to supply a username for a host
+# name or a full URL path (if the `credential.useHttpPath` config is set). For
+# sites like like `github.com`, you may want to use a different username for
+# different sub-paths, as the first path component designates an organisation.
+# If you have a work and personal GitHub identity, you may want to select that
+# identity based on GitHub organisation. It is not possible to do that directly
+# with the configuration used by `gitcredentials(7)`.
+#
+# This script operates as a git credential helper to sit between git and the
+# real credential helper you want to use (called the chained helper in this
+# script). It takes the host and path and searches your git config for a
+# username by walking the path component by component.  If the full path
+# matches multiple config entries, the most specific one is selected. The
+# username from that configuration is passed to the chained helper so it can
+# provide the credentials for the correct username.
+#
+# This helper requires that the config setting `credential.useHttpPath` is
+# true, however it effectively disables that for the credential helper it
+# chains to, so that helper will not request or store a password per path
+# (which is rarely what you want - you want to be able to select the username
+# per path and the password per host/username pair).
+#
+# If you do need to have the chained helper see the path, you can set the
+# config option `credential.subpath.passPath` to true, but beware that it will
+# ask for you password for every repository even if it knows the host and
+# username for another repo.
+
+# Config command reference:
+# git config --global credential.helper subpath store
+# git config --global credential.useHttpPath true
+# git config --global credential.subpath.passPath true
+# git config --global credential.https://github.com/org.username octocat
+
+#-----------------------------------------------------------------------------
+usage() {
+  printf 'Usage: %s {<options>...} <chain> <get|store|erase>\n' "${0##*/}"
+  printf '  where <chain> is the credential helper to chain to.\n'
+  printf 'Available options:\n'
+  printf '  -d      print debugging input/output (warning will output password)\n'
+  printf '  -o arg  pass <arg> to the chained helper (multiple allowed)\n'
+}
+
+#-----------------------------------------------------------------------------
+main() {
+  set -euo pipefail
+  declare -A attrs
+  declare -a attr_order
+  declare -a chain_args
+  debug=false
+
+  parse_args "$@"
+
+  [[ "${cmd}" =~ (erase|get|store) ]] || { usage >&2; exit 1; }
+
+  read_attrs
+
+  if [[ "${cmd}" == 'get' ]]; then set_username_from_context; fi
```

Comment:
this should probably do this for set and erase too.

Reply:
I've done it for `erase` but not `set`. It does not make sense to change a username that has a password with it (as is the case with `set`). Also `set` will be using the username and password that worked and came from a previous get.

I'm not sure when `erase` is used in a git workflow, but presumably it might be run to erase any credentials stored for a given repository path. In that case, we do need to do username mapping.